test/test_corpus.cc

```diff
@@ -495,6 +495,129 @@ TEST_P(ExpectationTest, PerPhaseTest) { // NOLINT
+        }
+
+        expectation = test.expectations.find("parse-tree-json");
+        if (expectation != test.expectations.end()) {
+            got["parse-tree-json"].append(nodes->toJSON(*gs)).append("\n");
+        }
+
+        ast::ParsedFile file;
+        core::MutableContext ctx(*gs, core::Symbols::root());
+        file = testSerialize(*gs, ast::ParsedFile{ast::desugar::node2Tree(ctx, move(nodes)), f.file});
+
+        expectation = test.expectations.find("ast");
+        if (expectation != test.expectations.end()) {
+            got["ast"].append(file.tree->toString(*gs)).append("\n");
+        }
+
+        expectation = test.expectations.find("ast-raw");
+        if (expectation != test.expectations.end()) {
+            got["ast-raw"].append(file.tree->showRaw(*gs)).append("\n");
+        }
+
+        // DSL pass
+        file = testSerialize(*gs, ast::ParsedFile{dsl::DSL::run(ctx, move(file.tree)), file.file});
+
+        expectation = test.expectations.find("dsl-tree");
+        if (expectation != test.expectations.end()) {
+            got["dsl-tree"].append(file.tree->toString(*gs)).append("\n");
+        }
+
+        expectation = test.expectations.find("dsl-tree-raw");
+        if (expectation != test.expectations.end()) {
+            got["dsl-tree-raw"].append(file.tree->showRaw(*gs)).append("\n");
+        }
+
+        // local vars
+        file = testSerialize(*gs, local_vars::LocalVars::run(ctx, move(file)));
+
+        // namer
+        {
+            core::UnfreezeSymbolTable symbolTableAccess(*gs);
+            vector<ast::ParsedFile> vTmp;
+            vTmp.emplace_back(move(file));
+            vTmp = namer::Namer::run(ctx, move(vTmp));
+            file = testSerialize(*gs, move(vTmp[0]));
+        }
+
+        expectation = test.expectations.find("name-tree");
+        if (expectation != test.expectations.end()) {
+            got["name-tree"].append(file.tree->toString(*gs)).append("\n");
+        }
+
+        expectation = test.expectations.find("name-tree-raw");
+        if (expectation != test.expectations.end()) {
+            got["name-tree-raw"].append(file.tree->showRaw(*gs));
+        }
+
+        newTrees.emplace_back(move(file));
+    }
+
+    // resolver
+    trees = resolver::Resolver::runTreePasses(ctx, move(newTrees));
+    auto newErrors = errorQueue->drainAllErrors();
+    errors.insert(errors.end(), make_move_iterator(newErrors.begin()), make_move_iterator(newErrors.end()));
+
+    for (auto &resolvedTree : newTrees) {
+        expectation = test.expectations.find("resolve-tree");
+        if (expectation != test.expectations.end()) {
+            got["resolve-tree"].append(resolvedTree.tree->toString(*gs)).append("\n");
+        }
+
+        expectation = test.expectations.find("resolve-tree-raw");
+        if (expectation != test.expectations.end()) {
+            got["resolve-tree-raw"].append(resolvedTree.tree->showRaw(*gs)).append("\n");
+        }
+    }
+
+    for (auto &gotPhase : got) {
```

Comment:
We repeat this same logic in multiple places, right? Should we make a helper function?